File: implementations/dart/lib/types/castling.dart
/// Type-safe castling rights representation.

import 'piece.dart';

/// Represents castling rights for both sides.
class CastlingRights {
  final bool whiteKingside;
  final bool whiteQueenside;
  final bool blackKingside;
  final bool blackQueenside;
  
  const CastlingRights({
    required this.whiteKingside,
    required this.whiteQueenside,
    required this.blackKingside,
    required this.blackQueenside,
  });
  
  /// All castling rights available.
  static const all = CastlingRights(
    whiteKingside: true,
    whiteQueenside: true,
    blackKingside: true,
    blackQueenside: true,
  );
  
  /// No castling rights.
  static const none = CastlingRights(
    whiteKingside: false,
    whiteQueenside: false,
    blackKingside: false,
    blackQueenside: false,
  );
  
  /// Parse from FEN castling string (e.g., "KQkq").
  factory CastlingRights.fromFen(String fen) {
    if (fen == '-') return none;
    
    return CastlingRights(
      whiteKingside: fen.contains('K'),
      whiteQueenside: fen.contains('Q'),
      blackKingside: fen.contains('k'),
      blackQueenside: fen.contains('q'),
    );
  }
  
  /// Convert to FEN castling string.
  String toFen() {
    if (!whiteKingside && !whiteQueenside && !blackKingside && !blackQueenside) {
      return '-';
    }
    
    final buffer = StringBuffer();
    if (whiteKingside) buffer.write('K');
    if (whiteQueenside) buffer.write('Q');
    if (blackKingside) buffer.write('k');
    if (blackQueenside) buffer.write('q');
    return buffer.toString();
  }
  
  /// Check if kingside castling is available for a color.
  bool canCastleKingside(Color color) {
    return color is White ? whiteKingside : blackKingside;
  }
  
  /// Check if queenside castling is available for a color.
  bool canCastleQueenside(Color color) {
    return color is White ? whiteQueenside : blackQueenside;
  }
  
  /// Remove kingside castling rights for a color.
  CastlingRights removeKingside(Color color) {
    return color is White
        ? CastlingRights(
            whiteKingside: false,
            whiteQueenside: whiteQueenside,
            blackKingside: blackKingside,
            blackQueenside: blackQueenside,
          )
        : CastlingRights(
            whiteKingside: whiteKingside,
            whiteQueenside: whiteQueenside,
            blackKingside: false,
            blackQueenside: blackQueenside,
          );
  }
  
  /// Remove queenside castling rights for a color.
  CastlingRights removeQueenside(Color color) {
    return color is White
        ? CastlingRights(
            whiteKingside: whiteKingside,
            whiteQueenside: false,
            blackKingside: blackKingside,
            blackQueenside: blackQueenside,
          )
        : CastlingRights(
            whiteKingside: whiteKingside,
            whiteQueenside: whiteQueenside,
            blackKingside: blackKingside,
            blackQueenside: false,
          );
  }
  
  /// Remove all castling rights for a color.
  CastlingRights removeAll(Color color) {
    return color is White
        ? CastlingRights(
            whiteKingside: false,
            whiteQueenside: false,
            blackKingside: blackKingside,
            blackQueenside: blackQueenside,
          )
        : CastlingRights(
            whiteKingside: whiteKingside,
            whiteQueenside: whiteQueenside,
            blackKingside: false,
            blackQueenside: false,
          );
  }
  
  @override
  bool operator ==(Object other) {
    return other is CastlingRights &&
        whiteKingside == other.whiteKingside &&
        whiteQueenside == other.whiteQueenside &&
        blackKingside == other.blackKingside &&
        blackQueenside == other.blackQueenside;
  }
  
  @override
  int get hashCode => Object.hash(whiteKingside, whiteQueenside, blackKingside, blackQueenside);
  
  @override
  String toString() => toFen();
}

File: implementations/rust/src/types/castling.rs
/// Type-safe castling rights
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct CastlingRights {
    pub white_kingside: bool,
    pub white_queenside: bool,
    pub black_kingside: bool,
    pub black_queenside: bool,
}

impl CastlingRights {
    /// Create new castling rights with all rights available
    pub const fn new() -> Self {
        Self {
            white_kingside: true,
            white_queenside: true,
            black_kingside: true,
            black_queenside: true,
        }
    }

    /// Create castling rights with no rights available
    pub const fn none() -> Self {
        Self {
            white_kingside: false,
            white_queenside: false,
            black_kingside: false,
            black_queenside: false,
        }
    }

    /// Create custom castling rights
    pub const fn custom(
        white_kingside: bool,
        white_queenside: bool,
        black_kingside: bool,
        black_queenside: bool,
    ) -> Self {
        Self {
            white_kingside,
            white_queenside,
            black_kingside,
            black_queenside,
        }
    }

    /// Remove white kingside castling
    pub fn remove_white_kingside(&mut self) {
        self.white_kingside = false;
    }

    /// Remove white queenside castling
    pub fn remove_white_queenside(&mut self) {
        self.white_queenside = false;
    }

    /// Remove black kingside castling
    pub fn remove_black_kingside(&mut self) {
        self.black_kingside = false;
    }

    /// Remove black queenside castling
    pub fn remove_black_queenside(&mut self) {
        self.black_queenside = false;
    }

    /// Remove all white castling rights
    pub fn remove_white(&mut self) {
        self.white_kingside = false;
        self.white_queenside = false;
    }

    /// Remove all black castling rights
    pub fn remove_black(&mut self) {
        self.black_kingside = false;
        self.black_queenside = false;
    }
}

impl Default for CastlingRights {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_new_castling_rights() {
        let rights = CastlingRights::new();
        assert!(rights.white_kingside);
        assert!(rights.white_queenside);
        assert!(rights.black_kingside);
        assert!(rights.black_queenside);
    }

    #[test]
    fn test_none_castling_rights() {
        let rights = CastlingRights::none();
        assert!(!rights.white_kingside);
        assert!(!rights.white_queenside);
        assert!(!rights.black_kingside);
        assert!(!rights.black_queenside);
    }

    #[test]
    fn test_remove_rights() {
        let mut rights = CastlingRights::new();
        rights.remove_white_kingside();
        assert!(!rights.white_kingside);
        assert!(rights.white_queenside);
    }
}
